### ISSUE TITLE
NLP to PDDL 연구

### ISSUE THREAD

**Issue by Masterjun12** [Masterjun12/GCS-with-LLM#2]
### ## 📜 현재까지의 연구 동향 (Literature Survey)

NLP to PDDL 변환 연구는 자연어의 모호성을 해결하고, 이를 구조화된 기호 논리로 매핑하는 것을 목표로 하며, 다음과 같은 흐름으로 발전해 왔습니다.

1.  **초기 연구: 규칙 및 템플릿 기반 의미 분석 (Rule-based Semantic Parsing)**
    * 초기 접근법은 특정 단어(예: "가져와")를 특정 프레디케이트(`(fetch)`)에 매핑하는 정교한 문법 규칙과 템플릿을 수작업으로 만드는 데 중점을 두었습니다.
    * **장점:** 특정 도메인에서는 높은 정확도를 보입니다.
    * **한계:** 새로운 도메인이나 다른 표현 방식에 확장하기가 매우 어렵고, 자연어의 다양성에 대처하기 힘듭니다.

2.  **중기 연구: 통계적 기계 번역 및 전통적 머신러닝**
    * 이 문제는 자연어(Source Language)를 PDDL(Target Language)로 번역하는 **'기계 번역(Machine Translation)'** 문제로 치환될 수 있습니다.
    * CRF(Conditional Random Fields)나 SMT(Statistical Machine Translation) 같은 통계적 모델을 사용하여 단어와 구문이 PDDL의 어떤 요소(객체, 프레디케이트)에 해당할 확률이 높은지를 학습했습니다.
    * **장점:** 규칙 기반보다 유연하며 데이터로부터 패턴을 학습할 수 있습니다.
    * **한계:** 대규모의 정렬된 (자연어, PDDL) 병렬 코퍼스가 필요하며, 문맥 전체를 이해하는 데 한계가 있습니다.

3.  **최신 연구: 신경망 기계 번역 (NMT) 및 거대 언어 모델 (LLM) 활용** 🤖
    * **Seq2Seq 모델:** RNN, LSTM 기반의 Encoder-Decoder 모델(Seq2Seq)에 Attention 메커니즘을 적용하여 자연어 시퀀스를 PDDL 시퀀스로 변환하는 방식이 주류가 되었습니다. 문장의 전체적인 문맥을 더 잘 이해할 수 있게 되었습니다.
    * **Transformer 및 Pre-trained LLM:** 현재 가장 활발한 연구는 **Transformer** 아키텍처를 기반으로 한 거대 언어 모델(BERT, T5, GPT 등)을 활용하는 것입니다.
        * **Fine-tuning:** 사전 훈련된 언어 모델을 (자연어, PDDL) 데이터셋에 추가로 학습(미세 조정)시켜 특정 도메인의 변환 성능을 극대화합니다. T5와 같은 Text-to-Text 모델이 이 작업에 매우 적합합니다.
        * **In-context Learning (Few-shot Prompting):** GPT-3와 같은 초거대 모델에는 별도의 학습 없이, 프롬프트에 몇 개의 변환 예시만 넣어주면 새로운 자연어 명령을 PDDL로 곧바로 생성해내는 능력을 보여줍니다.

---

### ## 🧪 제안하는 실험 내용 (Proposed Experiment)

현재 연구 동향에 맞춰, 사전 훈련된 언어 모델을 활용한 실험을 다음과 같이 설계할 수 있습니다.

* **연구 가설:** 특정 도메인(예: 드론 배송)에서, 범용 언어 능력을 갖춘 사전 훈련된 LLM(T5 등)을 Fine-tuning 하는 것이 전통적인 Seq2Seq 모델보다 더 적은 데이터로 더 높은 PDDL 변환 정확도를 보일 것이다.

* **실험 모델:**
    * **베이스라인 모델:** Attention 메커니즘을 갖춘 **Seq2Seq (LSTM or GRU)** 모델.
    * **제안 모델:** 사전 훈련된 **T5 (Text-to-Text Transfer Transformer)** 모델. T5는 모든 NLP 문제를 텍스트 입/출력 문제로 변환하여 푸는 데 강력한 성능을 보입니다.

* **데이터셋:**
    * 실험을 위한 (자연어, PDDL) 병렬 코퍼스를 구축합니다. 예를 들어 '드론 배송' 도메인이라면 다음과 같은 데이터를 생성합니다.
        * **자연어:** "드론을 A 기지로 보내서 배터리를 충전시켜줘."
        * **PDDL Goal:** `(:goal (and (at drone1 base_a) (is_charged drone1)))`

* **실험 절차:**
    1.  데이터셋을 Training, Validation, Test 세트로 분할합니다.
    2.  베이스라인 모델(Seq2Seq)을 Training 데이터로 처음부터 학습시킵니다.
    3.  제안 모델(T5)을 동일한 Training 데이터로 Fine-tuning 합니다.
    4.  두 모델을 Test 데이터셋으로 추론(inference)을 수행하여 PDDL 문을 생성하게 합니다.
    5.  생성된 PDDL 문을 아래의 평가지표를 사용하여 정량적으로 비교 분석합니다.

---

### ## 📊 실험 평가지표 (Evaluation Metrics)

생성된 PDDL의 품질을 평가하는 것은 단순 문자열 비교 이상을 요구하며, 사실에 기반한 다음 지표들이 널리 사용됩니다.

1.  **정확 매칭 (Exact Match Accuracy)**
    * 생성된 PDDL 문자열이 정답(Ground Truth) PDDL 문자열과 **토씨 하나 틀리지 않고 정확히 일치**하는 비율입니다.
    * **장점:** 측정하기 가장 간단합니다.
    * **단점:** 매우 엄격한 지표입니다. `(and (at a) (at b))`와 `(and (at b) (at a))`는 논리적으로 동일하지만, 이 지표에서는 틀린 것으로 간주합니다.

2.  **BLEU 점수 (BLEU Score)**
    * 기계 번역 분야에서 유래한 지표로, 생성된 문장(PDDL)이 정답 문장과 얼마나 많은 단어/구문(n-gram)을 공유하는지 측정합니다.
    * **장점:** 완전 일치하지 않더라도 유사도를 측정할 수 있습니다.
    * **단점:** 문법적 구조나 논리적 의미의 타당성은 전혀 보장하지 못합니다. BLEU 점수가 높아도 PDDL 파서가 오류를 일으킬 수 있습니다.

3.  **구문적 타당성 (Syntactic Validity)** 🎯
    * 생성된 PDDL이 **문법적으로 올바른지**의 여부입니다.
    * **측정 방법:** 표준 PDDL 파서(VAL 등)를 통해 파싱(parsing)에 성공하는 비율을 측정합니다.
    * **중요성:** 이 지표는 생성된 PDDL이 최소한 '실행 가능한' 형식인지를 판단하는 필수적인 기준입니다.

4.  **계획 성공률 (Plan Success Rate)** 🏆
    * **가장 중요하고 실질적인 평가지표입니다.**
    * **측정 방법:**
        1.  생성된 PDDL Goal과 PDDL Domain 파일을 AI 플래너에 입력합니다.
        2.  플래너가 **성공적으로 계획을 찾아내는지**의 비율을 측정합니다.
        3.  더 나아가, 생성된 계획이 원래 자연어 명령의 **의도를 실제로 만족시키는지**를 평가할 수도 있습니다.
    * **의미:** 모델이 단순히 보기 좋은 PDDL을 만드는 것을 넘어, 실제로 문제를 해결할 수 있는 의미론적으로 올바른 목표를 생성했는지를 평가합니다.
